File: servicios/Cliente.java
/**
 * @api {get} /CLIENTES/  GET CLIENTE
 * @apiName Categoria 1
 * @apiGroup 3.- CLIENTE
 *
 *
 * @apiSuccess {Number} Id Id Unico del Cliente .
 * @apiSuccess {Number} Nit  Nit del Cliente.
 * @apiSuccess {String} Nombre  Nombre Del Cliente.
 * @apiSuccess {Number} Telefono  Telefono Del Cliente.
 * @apiSuccess {Number} Fecha  Fecha De Registro Del Cliente.
 * @apiSuccess {Number} IdUsuario  Llave foranea de Usuario.

 * @api {get} /Clientes 1.- GET CLIENTE 
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *      {
 *           "status": "success",
 *           "code": "200"
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *    {
 *      'status' 	=> 'error',
 *      'code'		=> 404,
 *      'message' 	=> 'Cliente no disponible'
 *     }
 */

 /**
 * @api {get} /clientes/:id  GET clientes Por ID
 * @apiName Categoria 2
 * @apiGroup 3.- CLIENTE
 *

 * @apiParam {Number} Id Id Unico del Cliente .
 * @apiSuccess {Number} Id Id del Cliente.
 * @apiSuccess {Number} Nit  Nit del Cliente.
 * @apiSuccess {String} Nombre  Nombre Del Cliente.
 * @apiSuccess {Number} Telefono  Telefono Del Cliente.
 * @apiSuccess {Number} Fecha  Fecha De Registro Del Cliente.
 * @apiSuccess {Number} IdUsuario  Llave foranea de Usuario.
 *
 * @api {get} /clientes/:id 2.- GET CLIENTES POR ID
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *       {
 *           "status": "success",
 *           "code": "200"
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *      'status' 	=> 'error',
 *      'code'		=> 404,
 *      'message' 	=> 'Cliente no disponible'
 *     }
 */

/**
 * @api {post} /categoria/ POST INSERTAR CLIENTE

 * @apiName Categoria 3
 * @apiGroup 3.- CLIENTE
 
 * @apiSuccess {Number} Id Id del Cliente.
 * @apiSuccess {Number} Nit  Nit del Cliente.
 * @apiSuccess {String} Nombre  Nombre Del Cliente.
 * @apiSuccess {Number} Telefono  Telefono Del Cliente.
 * @apiSuccess {Number} Fecha  Fecha De Registro Del Cliente.
 * @apiSuccess {Number} IdUsuario  Llave foranea de Usuario.

 *
  * @api {post} /clientes 3.-POST INSERTAR USUARIO
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *      {
 *      'status' => 'success',
 *      'code'	 => 200,
 *      'message' => 'Cliente creada correctamente'
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *      'status' => 'error',
 *      'code'	 => 404,
 *      'message' => 'El Cliente NO se ha creado'
 *     }
 */

 /**
 * @api {put} /Clientes/:id  PUT ACTUALIZAR CLIENTES POR ID
 * @apiName Categoria 4
 * @apiGroup 3.- CLIENTE
 *
 * @apiParam {Number} Id Id Unico del Cliente .
 * @apiSuccess {Number} Id Id del Cliente.
 * @apiSuccess {Number} Nit  Nit del Cliente.
 * @apiSuccess {String} Nombre  Nombre Del Cliente.
 * @apiSuccess {Number} Telefono  Telefono Del Cliente.
 * @apiSuccess {Number} Fecha  Fecha De Registro Del Cliente.
 * @apiSuccess {Number} IdUsuario  Llave foranea de Usuario.
 *
 * @api {put} /Clientes/:id  4.- PUT ACTUALIZAR CLIENTES POR ID
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *      {
 *       'status' 	=> 'success',
 *       'code'		=> 200,
 *       'message' 	=> 'Se ha actualizado correctamente!!'
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *       'status' 	=> 'error',
 *       'code'		=> 404,
 *       'message' 	=> 'El Cliente no se ha actualizado!!'
 *     }
 */


  /**
 * @api {delete} /clentes/:id  DELETE ELIMINAR clientes POR ID
 * @apiName Categoria 5
 * @apiGroup 3.- CLIENTE
 *
 * @apiParam {Number} id ID unico Cliente.
 *
 * @api {delete} /Clientes/:id  5.- DELETE ELIMINAR CLIENTES POR ID
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *      {
 *       'status' 	=> 'success',
 *       'code'		=> 200,
 *       'message' 	=> 'El cliente se ha eliminado correctamente!!'
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *       'status' 	=> 'error',
 *       'code'		=> 404,
 *       'message' 	=> 'El Cliente no se ha eliminado!!'
 *     }
 */

File: cliente.php
<?php

require_once 'vendor/autoload.php';

/**
 *
 * NubeAndo.COM - Educar es Compartir
 *
 * @author: Angel Céspedes Quiroz
 * @Usuario: acq1305
 * @Web: https://www.nubeando.com
 * @Facebook: https://www.facebook.com/acq1305
 * @Skype: acq1305
 * @Linkedin: https://bo.linkedin.com/in/acq1305
 *
 */

$app = new \Slim\Slim();

$db = new mysqli('localhost', 'root', '', 'farmaciacumbre');

// Configuración de cabeceras
header('Access-Control-Allow-Origin: *');
header("Access-Control-Allow-Headers: X-API-KEY, Origin, X-Requested-With, Content-Type, Accept, Access-Control-Request-Method");
header("Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE");
header("Allow: GET, POST, OPTIONS, PUT, DELETE");
header("Content-Type: application/json");

$method = $_SERVER['REQUEST_METHOD'];
if($method == "OPTIONS") {
    die();
}

/*$loginToken = function (){
    $app = \Slim\Slim::getInstance();
    $token = $app->request->headers('ApiKey');
    if($token=='1234567'){

    }else{
        throw new Exception("Usuario no valido");
    }
}*/

// LISTAR TODOS DE USUARIOS
$app->get('/clientes', function() use($db, $app){
    $sql = 'SELECT * FROM cliente ORDER BY id DESC;';
    $query = $db->query($sql);

    $usuarios = array();
    while ($usuario = $query->fetch_assoc()) {
        $usuarios[] = $usuario;
    }

    $result = array(
        'status' => 'success',
        'code'	 => 200,
        'data' => $usuarios
    );

    echo json_encode($result);
});



$app->get('/clientes/:id', function($id) use($db, $app){
    $sql = 'SELECT * FROM cliente WHERE id = '.$id;
    $query = $db->query($sql);

    $result = array(
        'status'    => 'error',
        'code'      => 404,
        'message'   => 'Producto no disponible'
    );

    if($query->num_rows == 1){
        $producto = $query->fetch_assoc();

        $result = array(
            'status'    => 'success',
            'code'      => 200,
            'data'  => $producto
        );
    }

    echo json_encode($result);
});


// GUARDAR USUARIO
$app->post('/clientes', function() use($app, $db){
    $result = array(
        'status' => 'error',
        'code'   => 404,
        'message' => 'Producto NO se ha creado'
    );

    // $token = $app->request->headers('ApiKey');

    // if($token=='1234567'){

        $json = $app->request->getBody('json');
        $data = json_decode($json, true);

       
        if(!isset($data['nit'])){
            $data['nit']=null;
        }

        if(!isset($data['nombre'])){
            $data['nombre']=null;
        }

        if(!isset($data['telefono'])){
            $data['telefono']=null;
        }

        if(!isset($data['fecha'])){
            $data['fecha']=null;
        }

        if(!isset($data['usuario_id'])){
            $data['usuario_id']=null;
        }

       



        $query = "INSERT INTO cliente VALUES(NULL,".
           
            "'{$data['nit']}',".
            "'{$data['nombre']}',".
            "'{$data['telefono']}',".
            "'{$data['fecha']}',".
            "'{$data['usuario_id']}'".
            ");";

        $insert = $db->query($query);

        if($insert){
            $result = array(
                'status' => 'success',
                'code'   => 200,
                'message' => 'Producto creado correctamente'
            );
        }
    

    echo json_encode($result);

});

// ACTUALIZAR UN PRODUCTO

$app->put('/clientes/:id', function($id) use($db, $app){
    $json = $app->request->getBody('json');
    $data = json_decode($json, true);

    $sql = "UPDATE cliente SET ".
            "nit = '{$data["nit"]}', ".
            "nombre = '{$data["nombre"]}', ".
            "telefono = '{$data["telefono"]}', ".
            "fecha = '{$data["fecha"]}', ".
            "usuario_id = '{$data["usuario_id"]}' WHERE id = {$id}";
          
            

    $query = $db->query($sql);

    if($query){
        $result = array(
            'status'    => 'success',
            'code'      => 200,
            'message'   => 'El producto se ha actualizado correctamente!!'
        );
    }else{
        $result = array(
            'status'    => 'error',
            'code'      => 404,
            'message'   => 'El producto no se ha actualizado!!'
        );
    }

    echo json_encode($result);

});


// ELIMINAR UN USUARIO

$app->delete('/clientes/:id', function($id) use($db, $app){
    $sql = 'DELETE FROM cliente WHERE id = '.$id;
    $query = $db->query($sql);

    if($query){
        $result = array(
            'status'    => 'success',
            'code'      => 200,
            'message'   => 'El producto se ha eliminado correctamente!!'
        );
    }else{
        $result = array(
            'status'    => 'error',
            'code'      => 404,
            'message'   => 'El producto no se ha eliminado!!'
        );
    }

    echo json_encode($result);
});

$app->run();